Heading block: add heading level checker

ZebulanStanphill opened the pull request:
## Description
Closes #10581 and #20870. Blocked by #29610.

Derived from #14889; big thanks to @Jackie6!. This PR adds a heading level checker to the Heading block to warn users when they select a heading level that is most likely incorrect. I've rebased/refactored the PR to work with the latest `master`, and I've addressed a lot of the feedback provided on that PR.

Note that this PR does not change anything in the native mobile app.

## Screenshots
![image](https://user-images.githubusercontent.com/19592990/82958187-7b332f80-9f7a-11ea-85e5-bfc33996909d.png)

**UPDATE:** it now looks like this:

![image](https://user-images.githubusercontent.com/19592990/87800209-b718a180-c813-11ea-9573-87f6314b0534.png)
![image](https://user-images.githubusercontent.com/19592990/87800330-e29b8c00-c813-11ea-838a-1a0b0240aaf3.png)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ZebulanStanphill committed:
Original implementation by @Jackie6 (Rebased + major fixes/tweaks.)

Derived from https://github.com/WordPress/gutenberg/pull/14889.

Rebase and major fixes by @ZebulanStanphill.

Co-authored-by: Jackie6 <Jackie6@users.noreply.github.com>

ZebulanStanphill commented:
Remaining tasks/questions:

1: I need someone to test that this works as expected with screen readers.

2: For non-screen-reader users, there is no immediate indication that a heading level is invalid. After selecting a heading level, the dropdown immediately closes. How do we improve this? I see 2 potential solutions:

- Add some kind of indicator to the toolbar button that the heading level is invalid. (Color the icon yellow, perhaps?)
- Move the warning to appear within the block content itself when the block is selected. This would definitely be noticeable... maybe even too noticeable?

3: Here's the toughest issue: how do we make this work properly with headings that are not inside Heading blocks? There are already many 3rd party heading blocks in various plugins, not to mention the fact that Custom HTML and Classic blocks can contain headings.

Solving this issue here should also solve it for the content structure tool. It could potentially also solve the [problem](https://github.com/WordPress/gutenberg/pull/21234#issuecomment-621881956) I've run into with `<!--nextpage-->` detection in #21234.

enriquesanchez commented:
hey @ZebulanStanphill

I too noticed how hard it is to see the warning message because of how the menu immediately closes. I'm not sure I have the right answer to this, but I think that your idea of a warning indicator might be worth exploring:

<img width="393" alt="Frame 10" src="https://user-images.githubusercontent.com/3276087/83076117-9c5b5500-a03a-11ea-9b0d-a9210fd24be7.png">

I was confused by:
> The selected heading level may be invalid. See the content structure tool for more info.

What is the content structure tool?

-----

I also tested with VoiceOver and Safari. I was only able to get the warning correctly announced once, but after that, I wasn't able to get it again. I'm not sure why this is happening. I even tried creating a new heading block and I could not get the warning properly announced again. Is it somehow setup to only fire once?

![2020-05-27 16-57-37 2020-05-27 17_00_37](https://user-images.githubusercontent.com/3276087/83076813-d4af6300-a03b-11ea-89f3-4125325f7afe.gif)

ZebulanStanphill commented:
This is the content structure tool:
![image](https://user-images.githubusercontent.com/19592990/83082697-bfd9cc00-a049-11ea-8810-fded38a5fd56.png)

I currently have the effect running the `speak` function set to only run again when either the chosen level changes or the validity of the heading level changes. I have it working this way because that's how it was in #14889. How is it _supposed_ to work, in your opinion?

Also, while double-checking the code in the PR, I spotted some potential optimizations, so I applied them and rebased the PR. The code is now slightly cleaner.

enriquesanchez commented:
>This is the content structure tool:
Ah, I forgot how it was called! 🤦 

>I currently have the effect running the speak function set to only run again when either the chosen level changes or the validity of the heading level changes. I have it working this way because that's how it was in #14889. How is it supposed to work, in your opinion?

I was not aware of that. I think that because the warning is always there, so you see it every time you open the headings menu, I was expecting the `speak` function to follow a similar pattern and also be announced every time the menu was interacted with. But I don't have strong opinions about this, if this is how it was set up before, then I'm fine if it stays that way.

ZebulanStanphill commented:
I'm not actually sure how I should go about implementing a yellow dot-thing for the button. Any recommendations?

ZebulanStanphill commented:
Alright, I decided to try adding the dot in a similar way to how the color indicator for the RichText inline text color format tool works. I'm not entirely satisfied with how it looks (and I'd like feedback on any better way to approach this), but it does work:

![image](https://user-images.githubusercontent.com/19592990/83528349-9dfba180-a4ae-11ea-8ad0-9a2897b2e38d.png)

shaunandrews commented:
I honestly never knew that there was such a thing as an "incorrect heading level." What are the benefits to enforcing the "correct" heading level?

I realize we already do it in the Content Structure UI, but exposing it more widely — especially with an ever present _dot_ in the toolbar — seems like it could be more annoying than helpful.

Irregardless of all that, I think the warning message needs to be much easier to understand; Right now its full of jargon and phrases that will make little sense to most people. Referring users to seek out and find another UI isn't really helpful — the message should stand on it's own.

shaunandrews commented:
I think the warning should maybe refer to _why_ heading should be in order. Something like this might be more informative and helpful:

> Placing headings in the wrong order could make it difficult for visitors to use your site.

ZebulanStanphill commented:
@shaunandrews [Here's a pretty good explanation of semantic headings.](https://webaim.org/techniques/semanticstructure/#headings)  And here's [another](https://www.w3.org/WAI/tutorials/page-structure/headings/). Using the correct heading levels is important for both accessibility and search engine optimization (SEO).

I agree that it would be helpful to expand upon why the invalid headings can be an issue. I've updated the notice to include a brief description of why appropriate heading levels are important, and I've linked that sentence to this tutorial. (We already link to another tutorial on the same site in the Image block "Alt text" help text.)

As for referencing the content structure tool, I did that because that's how it was in the previous PR. @afercia [suggested](https://github.com/WordPress/gutenberg/pull/14889#issuecomment-481639813) that the warning be kept short and that additional info be provided in the content structure tool.

Has your opinion on this changed since then, @afercia? Also, how do you think this warning should be presented? And do you spot any issues with the current screen reader support?

ZebulanStanphill commented:
For reference, here's what the warning looks like right now:
![image](https://user-images.githubusercontent.com/19592990/83698294-b828a300-a5c6-11ea-8929-c8c1c8483791.png)

pablohoneyhoney commented:
Fascinating issue! 

Are there other instances where this dot makes sense? 
My concern is introducing new items within the toolbar that are solving only one-off cases.

From a UI design standpoint, the size of the dot itself is too heavy and unbalanced with the other elements surrounding it, which also creates visual tension for the lack of spacing between the H and the border of the toolbar.

ZebulanStanphill commented:
@pablohoneyhoney I definitely agree the dot is kind of weird. But I haven't got any better ideas.

Essentially, I'm trying to solve a new issue here: a way to indicate that attention should be given to a dropdown by somehow highlighting the button that opens the dropdown. Whatever approach we use could also be applied to the content structure tool, and possibly also plugin sidebar toggles. (E.g. the Yoast plugin icon could have an indicator if there are serious SEO problems detected, or something like that.)

On the other hand, maybe showing an indicator on the button isn't the right approach in the first place. But what other way is there to show that there's an issue? Putting a warning notice within the selected block content seems heavy-handed.

Additionally, putting it in the inspector is not an option, because that would hide it away even more. The block UI was designed around the idea that the most important stuff is either in the block content itself or in the block toolbar. Important stuff should never be in the inspector.

shaunandrews commented:
> Are there other instances where this dot makes sense?

Right now, the only other instance I can think of is color contrast warnings like in the Navigation or Group block when you change the text and background color. However, as @ZebulanStanphill mentions, there's also room for plugins to add warnings and I imagine whatever solution we come up with will also be used by these plugins.

<img width="382" alt="image" src="https://user-images.githubusercontent.com/191598/83772096-fff40c80-a650-11ea-8659-1615118dc5cd.png">

One concern I have is that we start adding dots for all sorts of warnings, and then we end up with a dozen dots scattered around the screen and they lose all meaning.

In thinking about this, I wonder if instead of highlighting the warning in the moment, we instead collect and highlight warnings when the document is updated or published. Something in the pre-publish panel, similar to FSE's multi-entity saving.

jasmussen commented:
The principle of this PR seems good. But I would agree with Pablo that we need to balance the desire to help the user pick the right heading level, with not weighing down the UI. 

In that vein, I'd remove the dot entirely — by default you insert H2s (because presumably the H1 is used for the site title). Which means if you intend to insert an incorrect heading level, you have to first insert a correct one and then explicitly change it to an incorrect one.

This flow allows us to drastically reduce the UI, so that any warnings are shown only in the context of the wrong action taken, which will likely help you understand why it's wrong. Here, you should not select the H1 level, so it's grayed:

![Untitled-1](https://user-images.githubusercontent.com/1204802/83845632-91bf5080-a709-11ea-9ce5-c0f0f119c2ff.png)

If you do select it, that's when we show a warning:

![Untitled-2](https://user-images.githubusercontent.com/1204802/83845642-971c9b00-a709-11ea-9942-0e40759f8963.png)

Whether "graying" is the right tool to indicate incorrect heading level, I'm not sure — I think it works. But I would agree the dot is not the right one to use, because the dot is meant to attract attention, rather than warn.

ZebulanStanphill commented:
@jasmussen That does seem like a better solution than the dot. (And I'll start working on updating the PR to use that approach now.) The only thing that concerns me is whether this graying-out has any accessibility downsides. Pinging @afercia, since he's an expert on this sort of thing.

jasmussen commented:
We can explore various other treatments for "don't pick this heading level" than graying, including showing the icon with an outline, or with a strike through, or with an explanation mark or anything else we find that strikes a good balance.

shaunandrews commented:
One question I have about making incorrect heading options look disabled, is that wouldn't all but on or two options be marked as incorrect?

ZebulanStanphill commented:
@shaunandrews It depends on what the preceding heading is. If the preceding heading was an `h4`, then anything from `h2` through `h5` would be valid. If the preceding heading was an `h2`, only `h2` or `h3` would be valid options.

shaunandrews commented:
So on a blank document (or one without any Headings) everything but H2 would be invalid. It feels weird in my head, so I'll wait for some working code to make a judgement. Thanks for all the great discussion above.

enriquesanchez commented:
I think @jasmussen's idea is a great one and agree it's best if we avoid the potential for cluterring the block's UI.

jasmussen commented:
> So on a blank document (or one without any Headings) everything but H2 would be invalid. It feels weird in my head, so I'll wait for some working code to make a judgement. Thanks for all the great discussion above.

I think with the right design for the discouraged heading levels, this will be fine. I can take a look monday, if the graying feels too disabled, we can look at an icon overlay.

afercia commented:
> by default you insert H2s (because presumably the H1 is used for the site title). Which means if you intend to insert an incorrect heading level, you have to first insert a correct one and then explicitly change it to an incorrect one.

> This flow allows us to drastically reduce the UI, so that any warnings are shown only in the context of the wrong action taken, which will likely help you understand why it's wrong. Here, you should not select the H1 level, so it's grayed:

This assumption is not correct. There are legitimate cases where a web page may _not_ have a default H1 heading. Nothing new here, these are known issues that have been discussed at length in the last years. Also, whether a page has a default H1 is ultimately a responsibility of the theme, a responsibility that the block editor should respect.

Just as an example, Twenty Thirteen provides different templates for its post formats and some of them don't have a H1: aside, status, quote. In the same way, other themes may provide special templates for special use cases where a H1 is not necessary. Admittedly, these would be edge cases but still valid cases.

Regarding the visual feedback:

![Untitled-2](https://user-images.githubusercontent.com/1204802/83845642-971c9b00-a709-11ea-9942-0e40759f8963.png)

I'm not sure how that text could be rendered within the headings toolbar. That's an ARIA toolbar and should exclusively contain interactive controls that belong to the toolbar. Additionally, there should be a non-visual way to communicate the incorrect heading level, for example an audible message.

> Whether "graying" is the right tool to indicate incorrect heading level, I'm not sure — I think it works. 

No objections to "graying" a button as long as the color contrast is sufficient. These are not disabled buttons, they are still actionable thus they must have a sufficient color contrast. Additionally, there should be an accessible way to communicate the "grayed" button indicates a not recommended level. This information can't rely on color alone so there's the need of a _shape_  for sighted users and some semantics or visually hidden text for non-sighted users.

ZebulanStanphill commented:
>I'm not sure how that text could be rendered within the headings toolbar. That's an ARIA toolbar and should exclusively contain interactive controls that belong to the toolbar. Additionally, there should be a non-visual way to communicate the incorrect heading level, for example an audible message.

Currently, the (visual version) of the warning is inside a `<Dropdown>`, and the warning is adjacent to a `Toolbar` containing the heading level options. (Note that this is different from the collapsed text alignment and block "alignment" controls where the options are all in a `<DropdownMenu>`.)

The text of the visible warning notice may not be readable by a screen reader, but that should be okay because a (plain text) version of the warning message is already announced to screen reader users upon opening the heading level toolbar and/or changing the heading level. So for screen-reader users, this feature could be considered complete.

It seems the major remaining design issue is figuring out how to use shapes to visually communicate that certain heading levels are not recommended.

The only decent idea that comes to mind is to use a squiggly line similar to the ones commonly used for spell-checking. What do you think?

jasmussen commented:
Idea: we don't change or gray any of the heading level icons at all, but simply show the warning upon selection. This seems the smallest possible step to test, to move things forward, and it doesn't preclude us coming back to it later and adding a squiggly line.

ZebulanStanphill commented:
@jasmussen The problem with that is that the dropdown immediately closes after selecting a heading level. The (non-screen-reader) user never gets to see the warning. Should the behavior of this dropdown be changed?

jasmussen commented:
Ah, that's an excellent aspect I somehow missed. There is precedence for changing this behavior, the global more menu does this. But I'm not sure it's a good idea here. 

We may need that squiggly line after all. I'll ponder a design.

jasmussen commented:
I took a stab a this today, and I ended back up at graying, but here are the in-between steps anyway:

<img width="556" alt="Screenshot 2020-06-09 at 09 47 47" src="https://user-images.githubusercontent.com/1204802/84122297-838a7080-aa38-11ea-85b6-39c4e9952358.png">

<img width="551" alt="Screenshot 2020-06-09 at 09 47 50" src="https://user-images.githubusercontent.com/1204802/84122309-89805180-aa38-11ea-9840-c673aade8c15.png">

<img width="583" alt="Screenshot 2020-06-09 at 09 57 04" src="https://user-images.githubusercontent.com/1204802/84122331-92712300-aa38-11ea-9fbc-0fd0d8fa2f06.png">

All of those feel too busy and confusing. Graying still feels like it works the best:

<img width="561" alt="Screenshot 2020-06-09 at 10 03 15" src="https://user-images.githubusercontent.com/1204802/84122386-a7e64d00-aa38-11ea-9e69-9f4be78a3276.png">

The gray there is #939393 which meets 3:0 contrast ratio for large text and UI objects.

enriquesanchez commented:
>should we keep the dropdown open after changing the heading level?

Maybe it should stay open if there's an error?

ZebulanStanphill commented:
That sounds like it could work. Anyone have any objections to that idea?

jasmussen commented:
If it's not hard to try, seems worth it. Ultimately it's best when these things open and close predicably and similarly, but the presence of a warning seems like an okay heuristic for why that might change.

ZebulanStanphill commented:
@jasmussen I spent some time trying to figure it out, but I wasn't able to figure it out. I currently don't even know what code is causing the dropdown to close in the first place. If I knew that, the solution might present itself pretty quickly.

jasmussen commented:
Took this for a quick spin:

![headings](https://user-images.githubusercontent.com/1204802/84654589-c2746680-af0f-11ea-9b02-d492efda1551.gif)

The icon stays grayed even in the block toolbar when it's the wrong heading level:

<img width="845" alt="Screenshot 2020-06-15 at 13 54 28" src="https://user-images.githubusercontent.com/1204802/84654652-dae48100-af0f-11ea-8bc4-036ff712af4a.png">

I can't tell if that's good or bad. I'm leaning towards that the color should be the same as other items, i.e. not grayed, and that we should solicit some help (for example from @gziolo perhaps?)  with making the menu stay open. The More menu has some code to make the menu stay open that we should be able to reuse here. 

But otherwise, just moving all heading levels to the block toolbar as opposed to the partial toolbar/sidebar split is a big win from this one. Nice work, thank you.

ZebulanStanphill committed:
Various improvements.

ZebulanStanphill commented:
I've changed the dropdown toggle in the toolbar to always be the standard black color. I took another look at the more menu code, but I still haven't figured out how it's preventing the close-on-click behavior.

ZebulanStanphill commented:
Something I've noticed that may be helpful to figuring this out: clicking the external link in the warning notice doesn't close the dropdown.

ZebulanStanphill commented:
Fixed failing tests. I still need help to figure out how to prevent the dropdown from auto-closing, so I'm marking this PR as blocked for now.

ntsekouras commented:
Hey @ZebulanStanphill , that's a great improvement here!

Regarding the blocker, I have made a branch from this one to try a solution here https://github.com/WordPress/gutenberg/tree/try/hover-heading-block-level-checker, with my commit changes being here https://github.com/WordPress/gutenberg/commit/0ac8d47c9d7f0105d0e191fe98aef13765242883 .

In my opinion a good solution would include:
- Inform the user if the selected header is `invalid or maybe invalid`
- Inform the user before selecting the `invalid or maybe invalid` heading

That led me to think that an `onHover` effect would be valuable here. In my commit I extracted the logic of showing the `HeadingLevelWarning` in the `Dropdown` and removed the flag if is `invalid` in the Warning. This would be even clearer that component always displays something, but is displayed conditionally.

Finally the `selectedLevel` in `HeadingLevelWarning` wasn't used before ( and not now yet ) but probably should be used to make the message more specific and the Warning message should have a bit more space on top, so as the `Tooltip` doesn't overlap with its content.

What do you think?

jasmussen commented:
Thanks for helping, @ntsekouras. This is what I see from your branch:

![subhead](https://user-images.githubusercontent.com/1204802/86111091-a1db0d80-bac6-11ea-836b-69a0b373484f.gif)

The idea here is good, but we need for the message to be visible also for keyboard-only usage, i.e. focus also. And we need to make sure that the message is also properly announced on focus. Both should be doable.

It does go in a different direction from what Zeb originally worked on, which was that the message showed up _after_ selecting it — this would require the menu to stay open even when clicking, as opposed to the default menu behavior of closing when clicking. This would be similar to the More menu, the ellipsis in the top right corner, which stays open when you select an item. 

I believe both approaches can work, and I have no strong preference for either. But if we were to address the specific issue of keeping the menu open, given Zeb has gotten the tests fixed up until that point, would you know how to help fix that? Thanks!

ntsekouras commented:
> we need for the message to be visible also for keyboard-only usage, i.e. focus also. And we need to make sure that the message is also properly announced on focus. Both should be doable.

That's easy to add yes with `onFocus` doing the same thing with `onHover`.

> which was that the message showed up after selecting it

I can't really understand why we should not be proactive and instead be reactive to a `not so good` heading choice, but I might have missed some context from the comments above..

> But if we were to address the specific issue of keeping the menu open, given Zeb has gotten the tests fixed up until that point, would you know how to help fix that?

This would be tricky indeed as the `selectedLevel` change triggers the component's render and the Popover is hidden on start. Even triggering a click to open after change, I think it would cause a flickering like effect.. I'm not really sure though.

jasmussen commented:
If the keep open approach is indeed hairy, seems like the hover and focus approach may be worth pivoting towards. What do you think, Zeb?

ZebulanStanphill commented:
Here are my concerns about the show-message-on-hover-and-focus approach:

- It could potentially be confusing to open up the level control, and see a warning because focus has automatically moved to the first option (the H1 option), even though the currently selected option is valid.
- Even when hovering/focusing a valid option, would we still show a warning because the currently-selected option is still invalid? That seems like it could get confusing, and it would take up even more text to explain that the hovered/focused option is valid but the currently-selected one is not.
- For screen readers, would we be announcing the entire warning message every time you move focus to another option? That's a lot of text to speak over and over again!
- It seems like this would make for some jumpy UI, with a huge chunk of text appearing and disappearing rapidly as you move your mouse.

I just came up with an alternative solution, which I have gone ahead and implemented. Let me know what you think:

- For potentially invalid options, we add something like "(may be invalid)" to the aria-label/tooltip. This conveys a much shorter warning to screen reader users, while also adding a bit more clarity for keyboard/mouse users in addition to the grey icons.
- We always show some help text below the heading level toolbar briefly explaining that you should use semantic heading levels.

![image](https://user-images.githubusercontent.com/19592990/86165447-e7a0d180-bad8-11ea-82ae-65e6c0c824bb.png)
![image](https://user-images.githubusercontent.com/19592990/86165488-fa1b0b00-bad8-11ea-96d7-a177ce5ffa2b.png)

jasmussen commented:
Thanks for keeping on this.

This is what I see:

![heading](https://user-images.githubusercontent.com/1204802/86214356-5d5d7980-bb7b-11ea-94fa-7780d676b3a2.gif)

The main downside is that it's quite a lot of text permanently visible there, which is likely to get tiring very fast after you've seen it once. For that reason I'd still lean towards an approach that showed this message only when you've actively picked the wrong heading level. 

Is there a way we can reduce the text so it feels less weighty? Given that you see it _every time_ you change headings, and that it links to a document going in-depth, perhaps we can reduce it quite a bit?

I don't know that I love this one:

<img width="575" alt="Screenshot 2020-07-01 at 09 26 52" src="https://user-images.githubusercontent.com/1204802/86215581-09539480-bb7d-11ea-9ac7-3d838da9d34e.png">

But it's definitely less weighty. At least something that fits on one line feels more appropriate, so that the information about picking the right level doesn't take up more space than the tools to pick them in the first place.

ntsekouras commented:
> For potentially invalid options, we add something like "(may be invalid)" to the aria-label/tooltip.

This is a good option as we are being proactive.

> We always show some help text below the heading level toolbar briefly explaining that you should use semantic heading levels.

We could also add an `info` icon or something similar to indicate that this is a helping text only? And I agree with @jasmussen about making the text shorter. Maybe something like `Learn how to use heading levels` ?

Finally we should check about the auto focus, if it can be set in the selected item, as in Chrome the tooltip stays there and by hovering with mouse to other headings, their tooltip overlaps with the first item's tooltip.

ntsekouras commented on packages/block-library/src/heading/heading-level-warning.js:
```diff
@@ -0,0 +1,42 @@
+);
+
+export default function HeadingLevelWarning( {
+	selectedLevel,
```
If the `selectedLevel` property will not be used eventually in the message, maybe we can remove it from the component.

ZebulanStanphill replied:
It's still being used to detect when the selected level has changed, so it can trigger a replay of the warning for screen reader users. Because of that, I don't think I can remove it as a dependency unless I extract the screen reader message out of the component. I think they should remain in the same component, however, since the message is essentially just the spoken version of the visual text.

ntsekouras commented on packages/block-library/src/heading/heading-level-dropdown.js:
```diff
@@ -67,36 +87,66 @@ export default function HeadingLevelDropdown( { selectedLevel, onChange } ) {
+							) }
+						</ExternalLink>
+					</p>
+					<HeadingLevelWarning
```
Maybe we can extract the display logic of `HeadingLevelWarning` here and remove the `selectedLevelMayBeInvalid` property from the component with something like this
```js
selectedLevelMayBeInvalid && <HeadingLevelWarning selectedLevel={ selectedLevel } />
```

This way it would be clear that the component always displays something, but is displayed conditionally

ZebulanStanphill replied:
Good idea. I'll do that.

ZebulanStanphill commented:
Alright, the message is now a lot shorter. I thought "Nest headings by their level" sounded a bit like a button to auto-fix the heading levels, so I went with something a bit different to make it clearer that it's a documentation link.

![image](https://user-images.githubusercontent.com/19592990/86267438-5428d880-bb8c-11ea-8649-744a663c9985.png)

enriquesanchez commented:
The shorter message feels a lot better. I did find some overlapping issues with the tooltips:

**Upon opening the dropdown, the tooltip covers the message below the heading options**
<img width="423" alt="Screen Shot 2020-07-01 at 17 25 45" src="https://user-images.githubusercontent.com/3276087/86297066-fca46000-bbbf-11ea-8022-d393499c2956.png">

**Hovering over other heading options does not close the previous tooltip:**
<img width="412" alt="Screen Shot 2020-07-01 at 17 26 02" src="https://user-images.githubusercontent.com/3276087/86297069-fe6e2380-bbbf-11ea-9b0b-104ac02ca4b9.png">

And we still have the issue of the dropdown closing immediately after selecting an incorrect heading level, so the user is not made aware of the error unless they open this menu again, right?

ZebulanStanphill commented:
@enriquesanchez
I've removed a style rule I added to remove space between the toolbar and help text. It now looks like this:
![image](https://user-images.githubusercontent.com/19592990/86304177-b1944800-bbd3-11ea-8bbb-bd82330a8c3c.png)

What do you think?

The tooltips overlapping other tooltips is not a problem with this branch. That's how it works everywhere when focus is moved using the keyboard or programmatically.

The help link, tooltip text, and greyed out icons are all designed to warn the user against using certain options before they select them. I was hoping this would make it unnecessary to prevent the dropdown from auto-closing. Implementing the auto-close behavior [would be difficult](https://github.com/WordPress/gutenberg/pull/22650#issuecomment-651698675), according to @ntsekouras.

ntsekouras commented:
> The help link, tooltip text, and greyed out icons are all designed to warn the user against using certain options before they select them. I was hoping this would make it unnecessary to prevent the dropdown from auto-closing.

I actually agree with the proactive approach and could work in this direction.

> Implementing the auto-close behavior would be difficult, according to @ntsekouras.

As I said that I believe so, but not really sure though, I've asked a couple of other devs to verify or not the above.

ZebulanStanphill commented:
Rebased to keep colors working after the changes in #23648.

enriquesanchez commented:
>I've removed a style rule I added to remove space between the toolbar and help text. It now looks like this:
image [...] What do you think?

That is much better. I still think things feel a bit tight in there... and I'm wondering, if there's always going to be a tooltip visible, maybe we can just add a bit more breathing room?

<img width="578" alt="Screen Shot 2020-07-03 at 16 39 28" src="https://user-images.githubusercontent.com/3276087/86498570-f3e09500-bd4b-11ea-93ec-a3cbb1134463.png">

ZebulanStanphill commented:
I've upped the margin on the help text to `$grid-unit-20` on all sides, which adds a bit more space. I feel like any more wouldn't feel right.
![image](https://user-images.githubusercontent.com/19592990/86501481-c0f3cc80-bd5e-11ea-9dd2-5a67e864b8fd.png)

shaunandrews commented:
Adding the extra spacing feels wrong. What about moving the warning above the buttons, or changing the tooltip to show above?

ZebulanStanphill commented:
If the link came before the toolbar, it would receive focus first, which I don't think is desirable. It would also be inconsistent with other help text.

I decided to go with the tooltips-above-the-buttons idea.

![image](https://user-images.githubusercontent.com/19592990/86516552-27bcc880-bde7-11ea-9149-d5b92d83b4cc.png)

I think this is good enough to ship. Any objections?

ZebulanStanphill commented:
Alright, it looks like everyone's happy with the latest iteration. All I need now is some approving reviews before I can merge.

Thank you all for your help with this PR; the tool is a lot more polished now than it would have been otherwise! :smile:

jasmussen reviewed (approved):
Thanks for the work, Zeb.

The interface is not the most elegant in the world. But it's also not something you see all the time, and it solves a real issue, and it does not preclude further improvements in the future. 

This PR could maybe use an additional sanity check, codewise, but otherwise, let's get this in and revisit if need be.

ZebulanStanphill commented:
Rebased to keep the PR fresh and updated the screenshots in the original post. All this PR needs is a final code review before merge.

jasmussen commented:
Still subscribed here, still appreciate your work. I hope as the dust settles people will come around to a code review here.

ntsekouras commented on packages/block-library/src/heading/heading-level-dropdown.js:
```diff
@@ -41,7 +46,19 @@ const POPOVER_PROPS = {
+	selectedLevel,
+	onChange,
+} ) {
+	const instanceId = useInstanceId( HeadingLevelDropdown );
```
You can provide the prefix you want in [useInstanceId](https://github.com/WordPress/gutenberg/blob/c1b05533dd1732a6f5d0bdd79080b1b671144b4a/packages/compose/src/hooks/use-instance-id/index.js#L25) and reuse it in both places below.

ex.
```
const instanceId = useInstanceId(
	HeadingLevelDropdown,
        'block-library-heading__heading-level-dropdown__help'
);
```

ZebulanStanphill replied:
But should I? While functionally, that would have the same effect, it seems a little confusing from a code readability perspective, since I'm prefixing an id referring to the entire component with a class name specific to one part of it.

ntsekouras replied:
The variable name could change to `descriptionId` to better reflect its content. That's why the second param exists. Of course it works as is now too..

ZebulanStanphill replied:
Oh, good point. It might also be slightly faster to do it the way you're suggesting, so I'll go ahead and make the change.

ntsekouras commented on packages/block-library/src/heading/use-heading-level-validator.js:
```diff
@@ -0,0 +1,99 @@
+	)[ 0 ];
+	const hasTitle = isTitleSupported && titleIsNotEmpty && titleNode;
+	const hasMultipleH1 = useMemo(
+		() => headings.filter( ( { level } ) => level === 1 ).length > 1,
```
You can use `some` here
```suggestion
		() => headings.some( ( { level } ) => level === 1 ),
```

ntsekouras commented on packages/block-library/src/heading/use-heading-level-validator.js:
```diff
@@ -0,0 +1,99 @@
+			targetLevel === 1 && hasTitle && ! hasMultipleH1;
+		const levelIsTooDeep = targetLevel > precedingLevel + 1;
+		const levelIsInvalid = levelIsDuplicateH1 || levelIsTooDeep;
+		const levelMayBeInvalid =
```
I think here we know when `levelIsInvalid` so `levelMayBeInvalid` should be just `levelAndPostTitleMayBothBeH1`, no?

Also you use only `levelMayBeInvalid` exported property in `HeadingLevelDropdown`, but since you know when it is invalid, you could use it to handle some options more specifically there.

It's certainly a big improvement already and not a blocker but maybe a follow-up PR could enhance it a bit more with more specific messages for `invalid` and `maybe invalid`.

ZebulanStanphill replied:
Yeah, I could go as far as explicitly handling each possible invalidity reason with a specific message, e.g. "The chosen heading level is too deep. Use a lower level or add a lower level heading before this one." But if I do something like that, then I'm not sure the "See the content structure tool for more info" bit should still be there anymore.

I'm not entirely sure how much info should be given in the warning. So for now I'll just keep it simple and as you suggested, explore improvements in a follow-up PR.

ntsekouras reviewed (approved):
Great work @ZebulanStanphill ! - I left some small comments but it's a great improvement 👍

Maybe we could add more specific titles in a follow-up PR, as I explain my reasoning in the comment?

ZebulanStanphill commented:
It's worth noting that this heading level validator only knows about core Heading blocks. It won't work with anything else. The proposed Accordion block (#21584) would use `h1`-`h6` elements (and not a Heading block), assuming we don't implement the accordion toggle and content as two different child blocks. We would have to update the logic in both blocks PLUS the content structure tool to ensure that the heading level validation logic is still accurate most of the time.

But even then, the logic will still be invalid when any 3rd party heading (or accordion) blocks are in use. I'm starting to question whether it's worth adding this if there's a good chance that it's warnings will often be wrong when 3rd party blocks are in use. And there's no way for plugins to fix this, as far as I am aware.

In my mind I keep coming back to #22874. I think we really need to have some kind of API that both core and 3rd party blocks can hook into for a centralized source of document outline info.

jasmussen commented:
As a separate conversation to have about the heading levels, it might be useful for developers to have access to a `Heading` component that automatically output the right heading level based on its context. This would not be a block, but simply something to import for blocks that bake in a heading.

gchtr commented:
> But even then, the logic will still be invalid when any 3rd party heading (or accordion) blocks are in use. I'm starting to question whether it's worth adding this if there's a good chance that it's warnings will often be wrong when 3rd party blocks are in use. And there's no way for plugins to fix this, as far as I am aware.

We disabled the `core/heading` block and use our own block for headings. Obviously, the The Document Outline is not showing up at all and the heading count is always at `0`. I think custom heading blocks should definitely be considered.

> In my mind I keep coming back to #22874. I think we really need to have some kind of API that both core and 3rd party blocks can hook into for a centralized source of document outline info.

Yes, please!

I wonder whether in addition to adding notifications about wrong heading levels and ways to fix them, a heading would also benefit from a **setting to adjust the display level**. Here’s what we currently use (probably there’s be a better way to design it):

![custom-heading](https://user-images.githubusercontent.com/2084481/90230127-59c73e80-de19-11ea-8fec-b745e9c9dc92.png)

I don’t know whether this was discussed before. In my experience, users often choose a heading based on its visual styles. When notifying them that they have the wrong heading level, there’s a conflict between heading styles and heading levels. Is it more important be semantically correct and have an accessible heading structure, or to have a visual appearance of the heading that best fits other elements on the page? I think that’s a question that we shouldn’t even have to ask ourselves. Heading level and appearance should be separated.

ZebulanStanphill commented:
The Heading block already has a "Typography" panel in the inspector containing font size options. (And that's a discussion for a separate issue anyway.)

ZebulanStanphill commented:
I've decided to mark this PR as blocked for now, due to the lack of support for 3rd party blocks. #22874 needs to be resolved before this feature can be merged, in my opinion.

aristath commented:
Is this one still blocked?

ZebulanStanphill commented:
@aristath Thanks for checking in. I was able to get #21234 merged without any new APIs, so I think I might be able to get this one working as well using a similar approach I've been wanting to attempt a revamp of this PR, but I haven't found the time yet. I'll comment again when I either get it working or determine that it's still blocked.

ZebulanStanphill committed:
Add support for headings outside of Heading blocks.

ZebulanStanphill commented:
@aristath Turns out it was indeed possible to get this PR fully working. The validator can now see all heading elements preceding the current block, not just core Heading blocks. This makes it a lot more accurate... in fact, it is now more accurate than the Document Outline tool in the editor's top bar, leading to the two disagreeing with each other.

Technically, the Document Outline is the buggy one, but since the notices on the Heading block tell you to read the Document Outline, we should probably fix its logic first. In the mean time, however, feel free to re-test and re-review this PR.

ZebulanStanphill commented:
I've created #29610 to fix the Document Outline.

masteradhoc commented:
@ntsekouras can we push this further? really like the solution by @ZebulanStanphill

ntsekouras commented:
>@ntsekouras can we push this further? really like the solution by @ZebulanStanphill

It seems there are still blocking issues here.. Can someone provide a TLDR for the current challenges of this?
